tools/eslint-plugin/dist/bin/no-internal-report.js

```diff
@@ -0,0 +1,48 @@
+#! /usr/bin/env node
+
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+"use strict"
+const path = require('path');
+const fs = require('fs');
+
+function getNodeModulesDir(startPath) {
+  let currPath = startPath;
+  let baseName = path.basename(currPath);
+  while (baseName !== "node_modules") {
+    currPath = path.resolve(currPath, "..")
+    if (currPath === "/")
+      return undefined;
+    baseName = path.basename(currPath);
+  }
+  return currPath;
+}
+
+const nodeModules = getNodeModulesDir(path.dirname(process.argv[1]));
+if (!nodeModules)
+  throw ("Could not find node_modules directory");
+
+const distDir = path.join(nodeModules, "@bentley/eslint-plugin/dist")
+if (!fs.existsSync(distDir))
+  throw ("Could not find required dir: " + distDir);
+
+// Run eslint with the appropriate configuration and formatter to get a report of the no-internal rule
+let args = [
+  "--no-eslintrc",
+  "-c", path.join(distDir, "configs/imodeljs-recommended.js"),
```

Comment:
this will still run all the rules in our recommended config, you can replace it with these options:
`--parser "@typescript-eslint/parser" --parser-options "{project:'tsconfig.json',sourceType:'module'}"`

Reply:
Thanks, I was wondering if that was possible. I'll give it a try and update the code.

Reply:
I don't get the same results when not using imodeljs-recommended.js. I'll email you the files as I can't seem to attach them here.

Reply:
@paulius-valiunas Any thoughts on my previous comment?

Reply:
sorry, I missed this one. I tried running your commands, and I think the quotes are off. I intentionally used double quotes so that I wouldn't have to escape the single quotes inside the json strings. Here's the commands I ran:
- `npx eslint --no-eslintrc -f visualstudio --plugin "@bentley" --parser @typescript-eslint/parser --parser-options "{project:'tsconfig.json',sourceType:'module'}" --rule "@bentley/no-internal:['error',{'tag':['internal','alpha','beta']}]" src/**/*.ts* > no_rec.txt`
- `npx eslint --no-eslintrc -c node_modules/@bentley/eslint-plugin/dist/configs/imodeljs-recommended.js -f visualstudio --plugin "@bentley" --parser @typescript-eslint/parser --parser-options "{project:'tsconfig.json',sourceType:'module'}" --rule "@bentley/no-internal:['error',{'tag':['internal','alpha','beta']}]" src/**/*.ts* > with_rec.txt`
In both cases I got 95 occurrences of `no-internal` in the package that I tested. The first command also included some "rule not found" noise, but other than that, I think they're identical. So maybe your shell is just not parsing the strings the way you intended?

Reply:
The problem seems to be related to specifying the `--parser` option. Here are two eslint commands using this file:

(I had to rename it in order to attach it here)
[RotateViewport.ts.txt](https://github.com/imodeljs/imodeljs/files/6581057/RotateViewport.ts.txt)

The first specifies `--parser`:

`npx eslint --no-eslintrc -c node_modules/@bentley/eslint-plugin/dist/configs/imodeljs-recommended.js -f visualstudio --plugin @bentley --parser @typescript-eslint/parser --rule "@bentley/no-internal:['error',{'tag':['internal','alpha','beta']}]" /Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts`

```
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(1,1): error @typescript-eslint/no-redeclare : Definition for rule '@typescript-eslint/no-redeclare' was not found.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(1,1): error @typescript-eslint/no-shadow : Definition for rule '@typescript-eslint/no-shadow' was not found.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.

5 problems
```

The second does not specify `--parser`:

`npx eslint --no-eslintrc -c node_modules/@bentley/eslint-plugin/dist/configs/imodeljs-recommended.js -f visualstudio --plugin @bentley --rule "@bentley/no-internal:['error',{'tag':['internal','alpha','beta']}]" /Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts`
```
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(1,1): error @typescript-eslint/no-redeclare : Definition for rule '@typescript-eslint/no-redeclare' was not found.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(1,1): error @typescript-eslint/no-shadow : Definition for rule '@typescript-eslint/no-shadow' was not found.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(20,32): error @bentley/no-internal : "HitBoxX" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(20,46): error @bentley/no-internal : "HitBoxY" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(20,60): error @bentley/no-internal : "HitBoxZ" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(21,33): error @bentley/no-internal : "HitBoxX" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(21,48): error @bentley/no-internal : "HitBoxY" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(21,62): error @bentley/no-internal : "HitBoxZ" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(22,32): error @bentley/no-internal : "HitBoxX" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(22,46): error @bentley/no-internal : "HitBoxY" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(22,60): error @bentley/no-internal : "HitBoxZ" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(23,33): error @bentley/no-internal : "HitBoxX" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(23,47): error @bentley/no-internal : "HitBoxY" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(23,62): error @bentley/no-internal : "HitBoxZ" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(24,34): error @bentley/no-internal : "HitBoxX" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(24,48): error @bentley/no-internal : "HitBoxY" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(24,62): error @bentley/no-internal : "HitBoxZ" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(25,31): error @bentley/no-internal : "HitBoxX" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(25,45): error @bentley/no-internal : "HitBoxY" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(25,59): error @bentley/no-internal : "HitBoxZ" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(91,5): error @bentley/no-internal : "BeUiEvent" is beta.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(91,5): error @bentley/no-internal : "BeUiEvent" is beta.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(91,5): error @bentley/no-internal : "BeUiEvent" is beta.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.

26 problems
```

Any idea what could be going on?

Reply:
Ok, it took a while for me to figure this out, but I think the problem is that you're providing both `-c` and `--parser` arguments. When you only use `-c`, you get both `parser` and `parserOptions` from the shared config. But when you added the `--parser` argument, that seems to have overridden both of them, so you also have to add `--parserOptions` to your command line to make it function properly.

And you should really remove the `-c`. The reason I suggested adding `--parser` and `--parserOptions` was to replace it so you could remove it. These two options are really all you need from the config, unless you want to also run other eslint rules - which you probably don't, since you're ignoring their output anyway.

Reply:
Thanks for investigating further. Unfortunately, the output isn't correct when run without the shared config.

`npx eslint --no-eslintrc -f visualstudio --plugin @bentley --parser @typescript-eslint/parser --parser-options '{project:'tsconfig.json',sourceType:'module'}' --rule '@bentley/no-internal: ['error', { 'tag': [ 'internal', 'alpha', 'beta' ]}]' src/mobile-ui/RotateViewport.ts`
```
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.
/Users/todd/macdev/field-r32/FieldApp/FieldUI/ModelWebApp/src/mobile-ui/RotateViewport.ts(92,5): error @bentley/no-internal : "ScreenViewport.animateFrustumChange" is internal.

3 problems
```

The HitBoxX/Y/Z uses should be reported but they aren't. So something in that shared config is causing more results to be found.

Reply:
can you try it with double quotes around parser-options? I just created a file containing this:
```ts
import { HitBoxX } from "@bentley/ui-components";

HitBoxX.Left;
```
and ran this command:
`npx eslint --no-eslintrc -f visualstudio --plugin "@bentley" --parser @typescript-eslint/parser --parser-options "{project:'tsconfig.json',sourceType:'module'}" --rule "@bentley/no-internal:['error',{'tag':['internal','alpha','beta']}]" src/test.ts`
and it worked:
`...src\test.ts(3,1): error @bentley/no-internal : enum "HitBoxX" is internal.`
I still think it's just the quotes, I don't see any other differences between what we're doing.

Reply:
It doesn't seem to be a problem with the quotes, could it be tsconfig.json related?

```
~/macdev/field-r32/FieldApp/FieldUI/modelwebapp$ cat src/HitPath.ts 
import { HitBoxX } from "@bentley/ui-components";

HitBoxX.Left;
~/macdev/field-r32/FieldApp/FieldUI/modelwebapp$ npx eslint --no-eslintrc -f visualstudio --plugin @bentley --parser @typescript-eslint/parser --parser-options "{project:'tsconfig.json',sourceType:'module'}" --rule "@bentley/no-internal: ['error', { 'tag': [ 'internal', 'alpha', 'beta' ]}]" src/HitPath.ts
no problems
```

Reply:
Maybe? I don't think it would change anything since the shared config contains the same parser-options as this command. I even verified with process monitor that both using the shared config and `--parser-options` refers to the same tsconfig.json file. The config also contains this, if that helps fix the problem...
```json
env: {
    "browser": true
  },
```
It could be replaced with `--env browser`.
The documentation says `env` is only for recognizing Javascript global variables, but who knows, maybe it can have some other effect too.

Reply:
I have committed a change so that only the no-internal rule will run.

Reply:
It was a version mismatch of the `typescript` dependency. I'll resolve this thread and change it to a peerDependency in the plugin. For now, the workaround is to use typescript ~4.1.0